### PR TITLE
adding cuda version of fisher_price example

### PR DESCRIPTION
**Pull request by WitekPokorski**
This pull request adds the CUDA version of the fisher_price example. The CUDA kernel consists of the 'process' (energy loss and pair production) run for each particle/step. For each iteration the 'main' schedules as many CUDA threads as particles in flight. New particle which are created during the pair-production process are pushed into the common BlockData container. Particles which die (E=0) are released from the container and the slots are used by new ones.

### CODE REVIEW

File: test/cufisher_price.cu [hunk 1]
```diff
@@ -2,12 +2,132 @@
...
+
+  // call the 'process'
+  if (r < 0.5f) {
+    // energy loss
+    float eloss = 0.2f * (*block)[particle_index].energy;
+    scor->totalEnergyLoss.fetch_add(eloss < 0.001f ? (*block)[particle_index].energy : eloss);
+    (*block)[particle_index].energy = (eloss < 0.001f ? 0.0f : ((*block)[particle_index].energy - eloss));
+
+    // if particle dies (E=0) release the slot
+    if ((*block)[particle_index].energy == 0) block->ReleaseElement(particle_index);
```
Comment (drbenmorgan) [inline test/cufisher_price.cu:57]:
```suggestion
    if ((*block)[particle_index].energy < 0.001f) block->ReleaseElement(particle_index);
```

I think this guarantees the release, given the cutoff above without worrying about exact floating point comparison.

File: test/cufisher_price.cu [hunk 2]
```diff
@@ -2,12 +2,132 @@
...
+  scor->totalEnergyLoss = 0;
+
+  // Allocate a block of tracks with capacity larger than the total number of spawned threads
+  // Note that if we want to allocate several consecutive block in a buffer, we have to use
+  // Block_t::SizeOfAlignAware rather than SizeOfInstance to get the space needed per block
+  using Block_t    = adept::BlockData<MyTrack>;
+  size_t blocksize = Block_t::SizeOfInstance(capacity);
+  char *buffer2    = nullptr;
+  cudaMallocManaged(&buffer2, blocksize);
+  auto block = adept::BlockData<MyTrack>::MakeInstanceAt(capacity, buffer2);
```
Comment (drbenmorgan) [inline test/cufisher_price.cu:108]:
```suggestion
  auto block = Block_t::MakeInstanceAt(capacity, buffer2);
```

**Review by drbenmorgan — CHANGES_REQUESTED** [review]

File: test/cufisher_price.cu [hunk 3]
```diff
@@ -2,12 +2,132 @@
...
+    // if particle dies (E=0) release the slot
+    if ((*block)[particle_index].energy == 0) block->ReleaseElement(particle_index);
+  } else {
+    // pair production
+    float eloss = 0.5f * (*block)[particle_index].energy;
+    (*block)[particle_index].energy -= eloss;
+
+    // here I need to create a new particle
+    auto secondary_track = block->NextElement();
+    if (!secondary_track) return;
```
Comment (agheata) [inline test/cufisher_price.cu:65]:
```suggestion
    assert (secondary_track != nullptr && "No slot available for secondary track" );
```
This is a bad situation and the program should crash, otherwise the energy loss will not add up to the primary energy

**Review by drbenmorgan — APPROVED** [review]
Fine for me! @agheata, if you have no further items, go ahead and merge!

**Review by agheata — APPROVED** [review]
